jme3-lwjgl3/src/main/java/com/jme3/system/lwjgl/LwjglWindow.java

```diff
@@ -0,0 +1,484 @@
+        if (!JmeContext.Type.Display.equals(type) && !JmeContext.Type.OffscreenSurface.equals(type) && !JmeContext.Type.Canvas.equals(type)) {
+            throw new IllegalArgumentException("Unsupported type '" + type.name() + "' provided");
+        }
+
+        this.type = type;
+    }
+
+    /**
+     * @return Type.Display or Type.Canvas
+     */
+    public JmeContext.Type getType() {
+        return type;
+    }
+
+    /**
+     * Set the title if its a windowed display
+     *
+     * @param title
+     */
+    public void setTitle(final String title) {
+        if (created.get() && window != -1) {
+            glfwSetWindowTitle(window, title);
+        }
+    }
+
+    /**
+     * Restart if its a windowed or full-screen display.
+     */
+    public void restart() {
+        if (created.get()) {
+            needRestart.set(true);
+        } else {
+            LOGGER.warning("Display is not created, cannot restart window.");
+        }
+    }
+
+    /**
+     * Apply the settings, changing resolution, etc.
+     *
+     * @param settings
+     */
+    protected void createContext(final AppSettings settings) {
+        glfwSetErrorCallback(errorCallback = new GLFWErrorCallback() {
+            @Override
+            public void invoke(int error, long description) {
+                final String message = Callbacks.errorCallbackDescriptionString(description);
+                listener.handleError(message, new Exception(message));
+            }
+        });
+
+        if (glfwInit() != GL_TRUE) {
+            throw new IllegalStateException("Unable to initialize GLFW");
+        }
+
+        glfwDefaultWindowHints();
+        glfwWindowHint(GLFW_VISIBLE, GL_FALSE);
+
+        // TODO: Add support for monitor selection
+        long monitor = NULL;
+
+        if (settings.isFullscreen()) {
+            monitor = glfwGetPrimaryMonitor();
+        }
+
+        final ByteBuffer videoMode = glfwGetVideoMode(glfwGetPrimaryMonitor());
+
+        if (settings.getWidth() <= 0 || settings.getHeight() <= 0) {
+            settings.setResolution(GLFWvidmode.width(videoMode), GLFWvidmode.height(videoMode));
+        }
+
+        window = glfwCreateWindow(settings.getWidth(), settings.getHeight(), settings.getTitle(), monitor, NULL);
+
+        if (window == NULL) {
+            throw new RuntimeException("Failed to create the GLFW window");
+        }
+
+        glfwWindowHint(GLFW_RESIZABLE, settings.isResizable() ? GL_TRUE : GL_FALSE);
+        glfwWindowHint(GLFW_DEPTH_BITS, settings.getDepthBits());
+        glfwWindowHint(GLFW_STENCIL_BITS, settings.getStencilBits());
+        glfwWindowHint(GLFW_SAMPLES, settings.getSamples());
+        glfwWindowHint(GLFW_STEREO, settings.useStereo3D() ? GL_TRUE : GL_FALSE);
+
+        int frameRateCap = settings.getFrameRate();
+
+        if (!autoFlush) {
+            frameRateCap = 20;
+        }
+
+        if (frameRateCap > 0) {
+            glfwWindowHint(GLFW_REFRESH_RATE, frameRateCap);
```

Comment:
This is not really correct. `GLFW_REFRESH_RATE` is the display refresh rate for fullscreen mode, whereas `settings.getFrameRate()` is a software-enforced limit of framerate. `GLFW_REFRESH_RATE` should be set to `settings.getFrequency()` instead.

Probably should stick to the LWJGL2 implementation for framerate.